services/middleware/src/transactions.rs

```diff
@@ -0,0 +1,192 @@
+// Copyright 2019 The Exonum Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Transaction logic for `MiddlewareService`.
+
+use exonum::runtime::{
+    rust::{CallContext, ChildAuthorization},
+    AnyTx, ExecutionError, InstanceStatus,
+};
+use exonum_derive::*;
+use exonum_proto::ProtobufConvert;
+use semver::VersionReq;
+use serde_derive::*;
+
+use crate::{proto, MiddlewareService};
+
+/// Errors of the `MiddlewareService`.
+#[derive(Debug, Clone, Copy, ExecutionFail)]
+pub enum Error {
+    /// The service instance targeted by the checked call does not exist.
+    NoService = 0,
+    /// The service instance targeted by the checked call is not active.
+    ServiceIsNotActive = 1,
+    /// The called service instance has an unexpected artifact.
+    ArtifactMismatch = 2,
+    /// The called service instance has an unsupported version.
+    VersionMismatch = 3,
+}
+
+/// Checked call to the service.
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, BinaryValue)]
+pub struct CheckedCall {
+    /// Expected name of the artifact.
+    pub artifact_name: String,
+    /// Version requirement(s) on the artifact.
+    pub artifact_version: VersionReq,
+    /// The call contents.
+    pub inner: AnyTx,
+}
+
+impl ProtobufConvert for CheckedCall {
```

Comment:
Can we implement `ProtobufConvert` for `VersionReq` and just derive it for `CheckedCall` then?

Reply:
Do you mean in the `exonum-proto` crate? Otherwise, the implementation is impossible as per Rust orphaning rules.

Reply:
Indeed. Then I think it doesn't worth it.

Reply:
But we may use this attribute to kill boiler-plate anyway.
https://github.com/witnet/protobuf-convert#overriding-conversion-rules